internal/app/upstream/client.go

```diff
@@ -55,11 +74,177 @@ type Response struct {
 //
 // The method does not block until the underlying connection is up.
 // Returns immediately and connecting the server happens in background
-// TODO: pass retry/timeout configurations
-func NewClient(ctx context.Context, url string) (Client, error) {
-	return &client{}, nil
+// TODO: pass retry configurations
+func NewClient(ctx context.Context, url string, callOptions CallOptions) (Client, error) {
+	conn, err := grpc.Dial(url, grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+
+	ldsClient := v2.NewListenerDiscoveryServiceClient(conn)
+	rdsClient := v2.NewRouteDiscoveryServiceClient(conn)
+	edsClient := v2.NewEndpointDiscoveryServiceClient(conn)
+	cdsClient := v2.NewClusterDiscoveryServiceClient(conn)
+
+	go shutDown(ctx, conn)
+
+	return &client{
+		ldsClient:   ldsClient,
+		rdsClient:   rdsClient,
+		edsClient:   edsClient,
+		cdsClient:   cdsClient,
+		callOptions: callOptions,
+	}, nil
+}
+
+// NewMockClient creates a mock implementation for testing
+func NewMockClient(
+	ctx context.Context,
+	ldsClient v2.ListenerDiscoveryServiceClient,
+	rdsClient v2.RouteDiscoveryServiceClient,
+	edsClient v2.EndpointDiscoveryServiceClient,
+	cdsClient v2.ClusterDiscoveryServiceClient,
+	callOptions CallOptions) Client {
+	return &client{
+		ldsClient:   ldsClient,
+		rdsClient:   rdsClient,
+		edsClient:   edsClient,
+		cdsClient:   cdsClient,
+		callOptions: callOptions,
+	}
+}
+
+func (m *client) OpenStream(ctx context.Context, request *v2.DiscoveryRequest) (<-chan *Response, error) {
+	var stream grpc.ClientStream
+	var err error
+	switch request.GetTypeUrl() {
+	case listenerTypeURL:
+		stream, err = m.ldsClient.StreamListeners(ctx)
+	case clusterTypeURL:
+		stream, err = m.cdsClient.StreamClusters(ctx)
+	case routeTypeURL:
+		stream, err = m.rdsClient.StreamRoutes(ctx)
+	case endpointTypeURL:
+		stream, err = m.edsClient.StreamEndpoints(ctx)
+	default:
+		return nil, fmt.Errorf("Unsupported request type")
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	signal := make(chan *version, 1)
+	signal <- &version{}
```

Comment:
Does xDS protocol require an empty version to start? Not clear why signal needs to buffered but response does not. It would be good to add a comment here.

Reply:
I intended both to be [unbufferred](https://golang.org/doc/effective_go.html#channels). I'll correct it.

From the [xds protocol](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack), the initial request is always empty.